### PR TITLE
7.x islandora 1872 - Allow custom value separator for Solr

### PR DESCRIPTION
**Pull request by BrandonMorr**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1872

* Other Relevant Links (Google Groups discussion, related pull requests, Release pull requests, etc.)

# What does this Pull Request do?

This will allow the user to specify a separator between multivalued fields in the search results.

# What's new?

* added new textfield in the admin form 'islandora_solr_search_field_value_separator'.
* added a check for empty values in the admin form validation handler. This is to prevent users from saving the form and having empty values passed to the display, causing results to look messy.

# How should this be tested?

* Ingest an object and create some multivalued fields (mods_name_personal_creator_namePart_ms for example).
* Navigate to your.site/admin/islandora/search/islandora_solr/settings and under the 'Default display settings' tab you will see a new field labelled 'Field value separator'.
* Enter test values, and also test that empty values are being set to NULL and falling back to default separators.
* Make sure your multivalued field is being displayed for search results.

# Interested parties
@Islandora/7-x-1-x-committers, @nhart

### CODE REVIEW

File: includes/utilities.inc [hunk 1]
```diff
@@ -342,7 +342,7 @@ function islandora_solr_prepare_solr_results($solr_results) {
           'url' => $object_result['object_url'],
           'options' => $options,
         );
-        $value = islandora_solr_truncate_field_display($value, $truncate_length[$field]['maximum_length'], $truncate_length[$field]['add_ellipsis'], $truncate_length[$field]['wordsafe'], $truncate_length[$field]['wordsafe_length'], "<br />", $link_options);
+        $value = islandora_solr_truncate_field_display($value, $truncate_length[$field]['maximum_length'], $truncate_length[$field]['add_ellipsis'], $truncate_length[$field]['wordsafe'], $truncate_length[$field]['wordsafe_length'], variable_get('islandora_solr_search_field_value_separator', '<br />'), $link_options);
```
Comment (whikloj) [inline includes/utilities.inc]: So here if you didn't specify a separator I'd get `<br />` (continued below)

File: includes/utilities.inc [hunk 2]
```diff
@@ -364,7 +364,7 @@ function islandora_solr_prepare_solr_results($solr_results) {
       }
 
       // Implode.
-      $value = is_array($value) ? implode(", ", $value) : $value;
+      $value = is_array($value) ? implode(variable_get('islandora_solr_search_field_value_separator', ', '), $value) : $value;
```
Comment (whikloj) [inline includes/utilities.inc]: But here I'd get `, `. So if someone sets this variable they can never go back to this default behaviour without removing the variable entirely from MySQL. Perhaps you actually need 2 variables?

File: includes/admin.inc [hunk 3]
```diff
@@ -408,6 +408,12 @@ function islandora_solr_admin_settings($form, &$form_state) {
     '#description' => t('Add navigation params to object links.'),
     '#default_value' => variable_get('islandora_solr_search_navigation', FALSE),
   );
+  $form['islandora_solr_tabs']['default_display_settings']['islandora_solr_search_field_value_separator'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Field value separator'),
+    '#description' => t('Characters to separate values in multivalued fields.'),
+    '#default_value' => variable_get('islandora_solr_search_field_value_separator', ''),
```
Comment (whikloj) [inline includes/admin.inc]: Related to the below comments about having 2 different displays run by a single choice, by default you don't show either of those defaults here.

Also, if we allow HTML to be added here and added to the display later. Should we be running a filter_xss or something on it?

File: includes/admin.inc [hunk 4]
```diff
@@ -1205,6 +1211,10 @@ function _islandora_solr_admin_settings_validate($form, &$form_state) {
   $limits = explode(',', $form_state['values']['islandora_solr_num_of_results_advanced']);
   $form_state['values']['islandora_solr_num_of_results_advanced'] = array_filter(array_map('trim', $limits), 'is_numeric');
   $form_state['values']['islandora_solr_namespace_restriction'] = preg_replace('/:$/', '', $form_state['values']['islandora_solr_namespace_restriction']);
+
+  if (empty($form_state['values']['islandora_solr_search_field_value_separator'])) {
+    $form_state['values']['islandora_solr_search_field_value_separator'] = NULL;
+  }
```
Comment (whikloj) [inline includes/admin.inc]: I'm not a huge stickler for this thing, but is the point here to avoid saving a bunch of spaces? If so, should that be done in the `_submit()` function and not in the `_validate()` function?

File: includes/utilities.inc (hunk 2 above)
Comment (whikloj) [reply includes/utilities.inc]: Actually as this is altering the returned values from Solr perhaps this shouldn't be altered at all and only the results output to the screen should be affected?

**Review by whikloj — CHANGES_REQUESTED** [review]
Some ramblings, I'll look some more tomorrow.

File: includes/admin.inc (hunk 4 above)
Comment (jordandukart) [reply includes/admin.inc]: Is the purpose of this to have the field be `#required` or `trim` away spaces or ?

File: includes/utilities.inc (hunk 1 above)
Comment (jordandukart) [reply includes/utilities.inc]: ++, different default to the `variable_get`.

File: includes/utilities.inc [hunk 5]
```diff
@@ -505,13 +505,13 @@ function islandora_solr_truncate_field_display($display_values, $max_length, $ad
 
     $updated_display_values
       = "<span class='toggle-wrapper'>" .
-      t("<span>!value !separator !show_more</span>", array(
+      t("<span>!value !show_more</span>", array(
         '!separator' => $separator,
```
Comment (jordandukart) [inline includes/utilities.inc]: Removed this variable but not the replacement?

File: includes/utilities.inc [hunk 6]
```diff
@@ -505,13 +505,13 @@ function islandora_solr_truncate_field_display($display_values, $max_length, $ad
...
         '!value' => $truncated_value,
         '!show_more' => ($value_is_truncated ? "<a href='#' class='toggler'>Show more</a>" : ""),
       ));
     if ($value_is_truncated) {
-      $updated_display_values .= t("<span>!original_value !separator<a href='#' class='toggler'>Show less</a></span>", array(
+      $updated_display_values .= t("<span>!original_value <a href='#' class='toggler'>Show less</a></span>", array(
         '!separator' => $separator,
```
Comment (jordandukart) [inline includes/utilities.inc]: Removed this variable but not the replacement?

File: includes/utilities.inc (hunk 2 above)
Comment (jordandukart) [reply includes/utilities.inc]: @whikloj++

File: includes/utilities.inc (hunk 5 above)
Comment (whikloj) [reply includes/utilities.inc]: @jordandukart is right but also you aren't changing how the `$truncated_value` is being created so why touch this at all?

File: includes/utilities.inc (hunk 6 above)
Comment (whikloj) [reply includes/utilities.inc]: I have the same question as above about whether this change is needed? I'll try and really test this out today or tomorrow, I might just be missing something.

File: includes/utilities.inc (hunk 2 above)
Comment (nhart) [reply includes/utilities.inc]: @whikloj if we implode here, we can't implode again at the display layer with a different separator.

File: includes/utilities.inc (hunk 5 above)
Comment (nhart) [reply includes/utilities.inc]: @jordandukart  good catch.
@whikloj this change isn't directly related to the new var, more of a bug related to separators. In the case of a comma separator, the last field value gets a comma appended to it. It's not actually separating values.

Eg: https://github.com/Islandora/islandora_xml_forms/blob/7.x/builder/transforms/mods_to_dc.xsl#L120 (terrible example, I know) if it's the last value, don't added a separator

Comment (whikloj) [reply includes/utilities.inc]: @nhart That makes more sense but currently I could only find it called using the `<br />` separator (on line [345](https://github.com/BrandonMorr/islandora_solr_search/blob/1361e18bad55ca19429311cb1a2bb40aabedb540/includes/utilities.inc#L345)).

Currently you get the **Show more** on a separate line. This will change the default display to have that **Show more** on the same line as the truncated item.

Perhaps that was the desired outcome, I just wasn't sure.

Comment (whikloj) [reply includes/utilities.inc]: Oh look at me be totally wrong. I missed the `variable_get` there. So I see what you mean. But do we want the **Show more** on its own line? Perhaps the separator can be removed and a `<br />` be hard coded in front of the **Show more** link?

Comment (nhart) [reply includes/utilities.inc]: yeah. I am ok with a hard coded br. Would be nice this was in a hook or in the theme layer so we all didn't have to fit into one format..but that is for another day :)

File: includes/utilities.inc [hunk 7]
```diff
@@ -304,6 +304,7 @@ function islandora_solr_prepare_solr_results($solr_results) {
 
     foreach ($solr_doc as $field => $original_value) {
       $value = (array) $original_value;
+      $separator = array_key_exists($field, $truncate_length) ? variable_get('islandora_solr_search_truncated_field_value_separator', "\n") : variable_get('islandora_solr_search_field_value_separator', ', ');
```
Comment (jordandukart) [inline includes/utilities.inc]: Think it makes more sense to use the `variable_get`s in the spots where they are used as opposed to some dynamic checking to set the separator as it's a little harder to read / follow without intricate knowledge.

File: includes/utilities.inc [hunk 8]
```diff
@@ -342,7 +342,7 @@ function islandora_solr_prepare_solr_results($solr_results) {
           'url' => $object_result['object_url'],
           'options' => $options,
         );
-        $value = islandora_solr_truncate_field_display($value, $truncate_length[$field]['maximum_length'], $truncate_length[$field]['add_ellipsis'], $truncate_length[$field]['wordsafe'], $truncate_length[$field]['wordsafe_length'], "<br />", $link_options);
+        $value = islandora_solr_truncate_field_display($value, $truncate_length[$field]['maximum_length'], $truncate_length[$field]['add_ellipsis'], $truncate_length[$field]['wordsafe'], $truncate_length[$field]['wordsafe_length'], $separator, $link_options);
```
Comment (jordandukart) [inline includes/utilities.inc:361]: Newline being replaced with a `br` or am I missing something?

File: includes/utilities.inc [hunk 9]
```diff
@@ -364,7 +364,7 @@ function islandora_solr_prepare_solr_results($solr_results) {
       }
 
       // Implode.
-      $value = is_array($value) ? implode(", ", $value) : $value;
+      $value = is_array($value) ? implode(filter_xss($separator, array('br')), $value) : $value;
```
Comment (jordandukart) [inline includes/utilities.inc]: If the separator is configurable does the `filter_xss` tags allowed need to be more broad than just `br`?

File: includes/utilities.inc [hunk 10]
```diff
@@ -448,6 +450,8 @@ function islandora_solr_islandora_basic_collection_backend_callable($collection_
  */
 function islandora_solr_truncate_field_display($display_values, $max_length, $add_ellipsis, $word_safe, $wordsafe_length, $separator, $link_options = NULL) {
   $updated_display_values = $display_values;
+  $separator = filter_xss($separator, array('br'));
```
Comment (jordandukart) [inline includes/utilities.inc]: Missed dis one?

File: includes/utilities.inc (hunk 8 above)
Comment (BrandonMorr) [reply includes/utilities.inc:361]: Fixed.

File: includes/utilities.inc [hunk 11]
```diff
@@ -447,7 +480,26 @@ function islandora_solr_islandora_basic_collection_backend_callable($collection_
...
+    'ul',
+    'ol',
+    'li',
+    'dl',
+    'dt',
+    'dd',
+    'br',
+    'span',
+    'div',
+  );
```
Comment (whikloj) [inline includes/utilities.inc]: If these tags and the ones defined in utilities.inc are the same, maybe defining them once as a `const` would be cleaner?

Comment (BrandonMorr) [reply includes/utilities.inc]: This can be done!

Comment (BrandonMorr) [reply includes/utilities.inc]: So because constant arrays weren't a thing until PHP 5.6, I create an islandora_solr_get_filter_tags function instead which returns the arrays. Hope this is an acceptable solution. Thanks for the feedback/review!

**Comment by BrandonMorr** [comment]
@whikloj this has passed internal review/testing, could you take another look please? Thanks!

**Review by whikloj — APPROVED** [review]

**Comment by nhart** [comment]
thanks @whikloj !
